internal/diffgenerator/diffgenerator.go

```diff
@@ -0,0 +1,197 @@
+		opt(cfg)
+	}
+	chunkReaders := []io.Reader{
+		bytes.NewBuffer([]byte(cfg.header)),
+	}
+	metadataChunkReader, err := makeMetadataChunkReader(cfg)
+	if err != nil {
+		return nil, err
+	}
+	chunkReaders = append(chunkReaders, metadataChunkReader)
+
+	for _, record := range cfg.records {
+		switch record.Type {
+		case RecordTypeUpdated:
+			updatedReader, newDataReader, err := makeUpdatedDataChunkReader(record.Offset, record.Length, record.Data)
+			if err != nil {
+				return nil, err
+			}
+			chunkReaders = append(chunkReaders, updatedReader, newDataReader)
+		case RecordTypeZero:
+			updatedReader, err := makeZeroDataChunkReader(record.Offset, record.Length)
+			if err != nil {
+				return nil, err
+			}
+			chunkReaders = append(chunkReaders, updatedReader)
+		}
+	}
+	chunkReaders = append(chunkReaders, bytes.NewBufferString("e"))
+	return &generatedDiffFileReader{
+		chunkReaders: chunkReaders,
+	}, nil
+}
+func makeUpdatedDataChunkReader(offset, length uint64, data []byte) (io.Reader, io.Reader, error) {
+	buf := bytes.Buffer{}
+	buf.WriteByte('w')
+
+	binary.Write(&buf, binary.LittleEndian, offset)
+	binary.Write(&buf, binary.LittleEndian, length)
+	chunkReader1 := bytes.NewBuffer(buf.Bytes())
+
+	if data == nil {
+		chunkReader2 := newRandomChunkReader(int(length))
+		return chunkReader1, chunkReader2, nil
+	}
+	if len(data) != int(length) {
+		return nil, nil, errors.New("data length does not match specified length")
+	}
+	chunkReader2 := bytes.NewBuffer(data)
+	return chunkReader1, chunkReader2, nil
+}
+
+func makeZeroDataChunkReader(offset, length uint64) (io.Reader, error) {
+	buf := bytes.Buffer{}
+	buf.WriteByte('z')
+
+	binary.Write(&buf, binary.LittleEndian, offset)
+	binary.Write(&buf, binary.LittleEndian, length)
+	chunkReader := bytes.NewBuffer(buf.Bytes())
+
+	return chunkReader, nil
+}
+
+func makeMetadataChunkReader(cfg *config) (io.Reader, error) {
+
+	buf := bytes.Buffer{}
+	if cfg.fromSnapName != "" {
+		buf.WriteByte('f')
+		err := binary.Write(&buf, binary.LittleEndian, int32(len(cfg.fromSnapName)))
+		if err != nil {
+			return nil, err
+		}
+		buf.WriteString(cfg.fromSnapName)
+	}
+
+	buf.WriteByte('t')
+	err := binary.Write(&buf, binary.LittleEndian, int32(len(cfg.toSnapName)))
+	if err != nil {
+		return nil, err
+	}
+	buf.WriteString(cfg.toSnapName)
+
+	buf.WriteByte('s')
+	err = binary.Write(&buf, binary.LittleEndian, int64(cfg.imageSize))
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes.NewBuffer(buf.Bytes()), nil
+}
+
+type generatedDiffFileReader struct {
+	chunkReaders []io.Reader
+}
+
+var _ io.Reader = &generatedDiffFileReader{}
+
+func (g *generatedDiffFileReader) Read(p []byte) (int, error) {
+	for i, chunkReader := range g.chunkReaders {
+		n, err := chunkReader.Read(p)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				continue
+			}
+			return 0, err
+		}
+		g.chunkReaders = g.chunkReaders[i:]
+		return n, nil
+	}
+	return 0, io.EOF
+}
+
+type randomChunkReader struct {
```

Comment:
Can we create an equivalent reader only with `rand.ChaCha8` and `io.LimitReader`?